Adding resource xPowerPlan (resolves issue #46)

johlju opened the pull request:
This is the PR that initiates the move of the xSQLServerPowerPlan resource to this module (see issue #46). This new xPowerPlan is based on the module xSQLServerPowerPlan but it is totally rewritten. and also improved somewhat. When this PR is merged I will continue the process to remove the old resource from the xSQLServer module.

This PR
- Removes the .gitignore file from this module.
- Adds the resource xPowerPlan
  - Also adds test for xPowerPlan, and an example.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xcomputermanagement/47)

<!-- Reviewable:end -->

msftclas commented:
Hi __@johlju__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

eshaparmar commented:
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8WsAjGpVct8BblBUd:-KS8WsAjGpVct8BblBUe:b9j9ng5) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L18)):*
> ```PowerShell
>     try
>     {
>         if ($Ensure -eq 'Absent' )
> ```

I am confused with $Ensure='Absent' scenario. If I want to ensure that highperformance is not the current active plan i.e it is absent then how will that work? 

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 21 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8R_K8tne9jqtfqPvi:-KS8R_K8tne9jqtfqPvj:b-u6odq1) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L21)):*
> ```PowerShell
>         {
>             # Hard-coded to plan name 'Balanced' because that is the default plan after OS installation
>             $Name = 'Balanced' 
> ```

Consider making 'Balanced' a script variable since it is being used at multiple places

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 34 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8V9nhmSG_xBpJs5Rr:-KS8V9nhmSG_xBpJs5Rs:b-dksnwh) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L34)):*
> ```PowerShell
>             else
>             {
>                 Write-Verbose "$Name is the not active plan"
> ```

if a plan is not active, then $activePlanName variable would not be set

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 34 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8VFXhViVRli6MgIMd:-KS8VFXhViVRli6MgIMe:b1pxcyd) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L34)):*
> ```PowerShell
>             else
>             {
>                 Write-Verbose "$Name is the not active plan"
> ```

Minor: Re-write the verbose message

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 47 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Vk1YCIUnu6TtulGu:-KS8Vk1YCIUnu6TtulGv:b-1esq2l) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L47)):*
> ```PowerShell
>     }
> 
>     $returnValue = @{
> ```

you could just return the hashtable directly, no need for its assignment to $returnValue

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 80 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8U-wAlGX6rv9C4HPE:-KS8U-wAlGX6rv9C4HPF:betx043) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L80)):*
> ```PowerShell
>         if ($PSCmdlet.ShouldProcess($Name, 'Activating power plan'))
>         {
>             $plan = Get-CimInstance -Name root\cimv2\power -Class Win32_PowerPlan -Filter "ElementName = '$Name'" 
> ```

Looks like you are relying on the error thrown by Get-CimInstance if a plan does not exist. This should be handled in the code and a proper error message should be thrown to the user that is more specific to the 'plan' that does not exist

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 112 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Xh_OKCyIydXLFhaI:-KS8Xh_OKCyIydXLFhaJ:b-gbfw9g) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L112)):*
> $result.Ensure -eq 'Absent' -and $result.Name -eq 'Balanced' )

This essentially means that 'Balanced' is absent. Not convinced if return value for absent makes sense. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Added a new context block in the test, to test so that the new $script:presentStateDefaultPlanName variable is working as expected.

---

Reviewed 4 of 6 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8WsAjGpVct8BblBUd:-KSChU7DJ2UAIM2aF2ts:b-j7u1xg) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L18)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> I am confused with $Ensure='Absent' scenario. If I want to ensure that highperformance is not the current active plan i.e it is absent then how will that work? 
</details>
I agree this is confusing (it's an inheritance from original resource in xSQLServer). 

When `Ensure = 'Absent'` it will always set the plan to 'Balanced' since that is the original state for a clean OS installation. 
I could remove the `Ensure` parameter altogether and just keep the plan name. Actually setting the plan to 'Balanced' to remove 'High performance'  would be more logical. 

I'm all for
* Remove `Ensure` parameter.
* Make $Name mandatory.
* Removing default value for $Name parameter.

What do you think? 

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 21 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8R_K8tne9jqtfqPvi:-KSCov64U_X6K_jxS8Wt:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L21)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> Consider making 'Balanced' a script variable since it is being used at multiple places
</details>
Done

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 34 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8V9nhmSG_xBpJs5Rr:-KSCowXIvWYzHH3LPYmF:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L34)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> if a plan is not active, then $activePlanName variable would not be set
</details>
Done

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 34 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8VFXhViVRli6MgIMd:-KSCozk6ChTQhuOCzI5v:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L34)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> Minor: Re-write the verbose message
</details>
Done

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 47 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Vk1YCIUnu6TtulGu:-KSCp0dQjtkI-GAYTfVG:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L47)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> you could just return the hashtable directly, no need for its assignment to $returnValue
</details>
Done

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 80 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8U-wAlGX6rv9C4HPE:-KSCp4YTILD3KAQKQT2-:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L80)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> Looks like you are relying on the error thrown by Get-CimInstance if a plan does not exist. This should be handled in the code and a proper error message should be thrown to the user that is more specific to the 'plan' that does not exist
</details>
Done

---

*[DSCResources/xPowerPlan/xPowerPlan.psm1, line 112 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Xh_OKCyIydXLFhaI:-KSCnAPrAgg9Lzn9J8sq:b-w1sxuc) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/xPowerPlan/xPowerPlan.psm1#L112)):*
<details><summary><i>Previously, eshaparmar (Esha Parmar) wrote…</i></summary>
> > $result.Ensure -eq 'Absent' -and $result.Name -eq 'Balanced' )
> 
> This essentially means that 'Balanced' is absent. Not convinced if return value for absent makes sense. 
</details>
Waiting for answer on first comment for this one. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KSCmjtd7rsuuOA_XzjE:b-n025ue)*
<!-- Sent from Reviewable.io -->

johlju commented on DSCResources/xPowerPlan/xPowerPlan.schema.mof:
```diff
@@ -0,0 +1,6 @@
+[ClassVersion("1.0.0.0"), FriendlyName("xPowerPlan")]
+class xPowerPlan : OMI_BaseResource
+{
+    [Key, Description("Use value 'Present' if the plan name should be activated. Use value 'Absent' to active the default plan for the operating system (hard-coded to plan 'Balanced')."), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
```
@TravisEz13 Will fix this during next week, when I'm back from Ignite. Thanks for reviewing!

johlju replied:
@TravisEz13 Requested changes done. If you would be so kind to review again whenever you have an opportunity

johlju commented:
Reviewed 7 of 8 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8WsAjGpVct8BblBUd:-KTPeF-gN81beYoWTePp:b-saetbm) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L18)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> I agree this is confusing (it's an inheritance from original resource in xSQLServer). 
> 
> When `Ensure = 'Absent'` it will always set the plan to 'Balanced' since that is the original state for a clean OS installation. 
> I could remove the `Ensure` parameter altogether and just keep the plan name. Actually setting the plan to 'Balanced' to remove 'High performance'  would be more logical. 
> 
> I'm all for
> * Remove `Ensure` parameter.
> * Make $Name mandatory.
> * Removing default value for $Name parameter.
> 
> What do you think? 
</details>
Change it to a single-instance resource

---

*[DSCResources/xPowerPlan/xPowerPlan.schema.mof, line 4 at r2](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KRxm5t4TzM-giV7ZlE7-r2-4:-KTPeQd98-FnMGY7YF76:bdz3jpx) ([raw file](https://github.com/powershell/xcomputermanagement/blob/98fed21fc48bbedd6771531fdc1bd4f40f3ef7df/DSCResources/xPowerPlan/xPowerPlan.schema.mof#L4)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> @TravisEz13 Will fix this during next week, when I'm back from Ignite. Thanks for reviewing!
</details>
Done. Sorry it took longer for me to start on this than I was aiming at. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
@TravisEz13 Requested changes done. If you would be so kind to review again whenever you have an opportunity

eshaparmar commented on DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1:
```diff
@@ -0,0 +1,108 @@
+        [Parameter(Mandatory = $true)]
+        [ValidateSet('Yes')]
+        [System.String]
+        $IsSingleInstance,
```
Doesnt look like you are doing anything with $IsSingleInstance. What is its purpose? Can you please add some comments? Thanks!

PlagueHO replied:
Hi @eshaparmar - this is the best practice for single instance resources. See this [MSDN page](https://msdn.microsoft.com/en-us/powershell/dsc/singleinstance). @TravisEz13 also wrote [this article](https://blogs.msdn.microsoft.com/powershell/2015/06/22/want-to-write-a-dsc-resource-where-only-a-single-instance-can-be-configured/) on it. :smile:

johlju replied:
@eshaparmar I could add # This is the best practice when writing a single-instance DSC resource. Would that suffice?

mgreenegit commented:
There is another PowerPlan resource in the PR below that is class based.  Just FYI if you would like to take a look at it @johlju.  It has not moved forward in some time.
https://github.com/torgro/PowerPlan/pull/1

johlju commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 10 at r3](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KTPcy2Z7mU44sbpsP92-r3-10:-KW8fMPQ8adyouiY4EpP:bs1pj14) ([raw file](https://github.com/powershell/xcomputermanagement/blob/74aebb6353c072fc2ffdd390da8514611ce08d0b/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L10)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary>
> Hi @eshaparmar - this is the best practice for single instance resources. See this [MSDN page](https://msdn.microsoft.com/en-us/powershell/dsc/singleinstance). @TravisEz13 also wrote [this article](https://blogs.msdn.microsoft.com/powershell/2015/06/22/want-to-write-a-dsc-resource-where-only-a-single-instance-can-be-configured/) on it. :smile:
</details>
@eshaparmar I could add `# This is the best practice when writing a single-instance DSC resource`. Would that suffice? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
@mgreenegit Looked over the code. I don't see that any of the logic is adding any value to this PR, it is basically the same. 

But I can take two things from your PR:
- Adding comment-based help
- Splatting the variables when calling `Get-CimInstance`

And after comparing the two versions I also see that I have added `SupportsShouldProcess` and `$PSCmdlet.ShouldProcess`. I will make sure to remove these.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KW8jvL_mJm52ZUU50F6:b-fedi2)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 8 files at r3, 1 of 3 files at r4.
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 10 at r3](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KTPcy2Z7mU44sbpsP92-r3-10:-KW9_2PsCII3MZdwHias:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/74aebb6353c072fc2ffdd390da8514611ce08d0b/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L10)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> @eshaparmar I could add `# This is the best practice when writing a single-instance DSC resource`. Would that suffice?
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
@kwirkykat and/or @mbreakey3 Would you please take a look at this PR and see if you can get it moving? If there is something I need to do let me know.

mbreakey3 commented:
Reviewed 1 of 7 files at r1, 4 of 8 files at r3, 2 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 112 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Xh_OKCyIydXLFhaI:-KWt0OXLKZc_kHHZkE3O:b6v1a5n) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L112)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Waiting for answer on first comment for this one. 
</details>
Maybe checking if it's active or not as well, and adding that as a param?

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 26 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWszWNhYQmcTMbj5Sxy:-KWszWNhYQmcTMbj5Sxz:b-1njp4w) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L26)):*
> ```PowerShell
>         $IsSingleInstance,
> 
>         [parameter(Mandatory = $true)]
> ```

'Parameter'

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 56 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt-ABrdD-GY01fbPoo:-KWt-ABrdD-GY01fbPop:b-c8a9ro) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L56)):*
> ```PowerShell
>         else
>         {
>             throw "Unable to find the power plan $Name." 
> ```

Any strings for errors or verbose statements should be in there own .strings file. Check out the format in in our [WindowsOptionalFeature](https://github.com/PowerShell/PSDscResources/blob/dev/DscResources/MSFT_WindowsOptionalFeature/en-US/MSFT_WindowsOptionalFeature.strings.psd1).

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 65 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt-_FRsS8XDkbPBtem:-KWt-_FRsS8XDkbPBten:b-jntg8e) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L65)):*
> ```PowerShell
> 
>     return @{
>         IsSingleInstance = $IsSingleInstance
> ```

Would it make more sense to return if the plan is active or not instead of $IsSingleInstance, since this variable will always be the same?

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 94 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt-uV7P0GiZWIplDhI:-KWt-uV7P0GiZWIplDhJ:b-1njp4w) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L94)):*
> ```PowerShell
>         $IsSingleInstance,
> 
>         [parameter(Mandatory = $true)]
> ```

'Parameter'

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 115 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt00iq-nC4S1qbGWlb:-KWt00iq-nC4S1qbGWlc:b-7rdb9) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L115)):*
> ```PowerShell
>     catch
>     {
>         Throw "Unable to set the power plan $Name to the active plan. Error message: $($_.Exception.Message)" 
> ```

This should be in the strings file as well.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 144 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0A6dqZySjyhhQ2ey:-KWt0A6dqZySjyhhQ2ez:b-1njp4w) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L144)):*
> ```PowerShell
>         $IsSingleInstance,
> 
>         [parameter(Mandatory = $true)]
> ```

'Parameter'

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0fvyxVNrrgLx5gIU:-KWt0fvyxVNrrgLx5gIV:bfvtwdz) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof#L1)):*
> ```
> [ClassVersion("1.0.0.0"), FriendlyName("xPowerPlan")]
> ```

single quotes throughout this doc.

---

*[Examples/Sample_xPowerPlan.ps1, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0qtvFOfll8ofqZmq:-KWt0qtvFOfll8ofqZmr:b-l8udhz) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Examples/Sample_xPowerPlan.ps1#L1)):*
> ```PowerShell
> configuration Sample_xPowerPlan
> ```

Can you add a description comment at the top about what this example does?

---

*[Examples/Sample_xPowerPlan.ps1, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0xpItcV90j0y8T82:-KWt0xpItcV90j0y8T83:b-ftish9) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Examples/Sample_xPowerPlan.ps1#L1)):*
> ```PowerShell
> configuration Sample_xPowerPlan
> ```

'Configuration'

---

*[Examples/Sample_xPowerPlan.ps1, line 5 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0nejTrtMYO1mIrLY:-KWt0nejTrtMYO1mIrLZ:b-re2ytq) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Examples/Sample_xPowerPlan.ps1#L5)):*
> ```PowerShell
>     param
>     (
>         [string[]] $NodeName = 'localhost'
> ```

'String[]'  

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 23 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt1ClQ25PD9tkIcW9o:-KWt1ClQ25PD9tkIcW9p:b-1gsryu) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L23)):*
> ```PowerShell
> #endregion HEADER
> 
> function Invoke-TestSetup {
> ```

you can remove this.

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 34 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt1FjaofyWHGKUZ-2a:-KWt1FjaofyWHGKUZ-2b:b-1kqndq) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L34)):*
> ```PowerShell
> {
> 
>     Invoke-TestSetup
> ```

remove this. And the extra newlines around it.

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 82 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt25jKUU9GYogLfcfG:-KWt25jKUU9GYogLfcfH:b-e24ann) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L82)):*
> ```PowerShell
> 
>             It 'Should throw an error' {
>                 { Get-TargetResource @testParameters } | Should Throw
> ```

this should be checking that the correct error is thrown.

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 131 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt2GOJCiuU0hAhwOzz:-KWt2GOJCiuU0hAhwP--:bfoju3z) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L131)):*
> ```PowerShell
> 
>             It 'Should catch the error thrown by Invoke-CimMethod' {
>                 { Set-TargetResource @testParameters } | Should Throw
> ```

again, should be checking that the correct error is thrown

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 112 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Xh_OKCyIydXLFhaI:-KWt5LK8ZWdomC309sZN:b-tj1zo8) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L112)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Maybe checking if it's active or not as well, and adding that as a param?
</details>
Nevermind, this is fine. I see what're checking for now. You should rename $result to $getTargetResourceResult though to be more descriptive. 

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 65 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt-_FRsS8XDkbPBtem:-KWt5B9iuub7qHytTeWS:bs9c2hd) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L65)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Would it make more sense to return if the plan is active or not instead of $IsSingleInstance, since this variable will always be the same?
</details>
Nevermind, I see what you're doing here. Disregard this message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 3 files at r4.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0fvyxVNrrgLx5gIU:-KWvkGRBfxr9NpjSl--D:bgc3ni5) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> single quotes throughout this doc.
</details>
I have to ask... Are you sure? :smile:
`Test-xDscSchema` does not like that. And neither does `mofcomp.exe`.

```
PS > Test-xDscSchema .\MSFT_xPowerPlan.schema.mof
.\MSFT_xPowerPlan.schema.mof (1): error SYNTAX 0X80044007: Illegal constant value. (Numeric value out of range or strings without quotes)
Compiler returned error 0x80044007
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Changed error-handling a bit when I localized to be able to localize all error messages.

---

Reviewed 1 of 8 files at r3, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 112 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Xh_OKCyIydXLFhaI:-KWvxkLHNMz7VtT171RV:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L112)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Nevermind, this is fine. I see what're checking for now. You should rename $result to $getTargetResourceResult though to be more descriptive. 
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 26 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWszWNhYQmcTMbj5Sxy:-KWvx9iWA45QTjIsQgYe:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L26)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> 'Parameter'
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 56 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt-ABrdD-GY01fbPoo:-KWvxBR5Cx2NqAMRkfWg:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L56)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Any strings for errors or verbose statements should be in there own .strings file. Check out the format in in our [WindowsOptionalFeature](https://github.com/PowerShell/PSDscResources/blob/dev/DscResources/MSFT_WindowsOptionalFeature/en-US/MSFT_WindowsOptionalFeature.strings.psd1).
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 94 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt-uV7P0GiZWIplDhI:-KWvxKWWTX97_kxwNOW1:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L94)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> 'Parameter'
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 115 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt00iq-nC4S1qbGWlb:-KWvxLgnp4nNgo9ehuqH:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L115)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> This should be in the strings file as well.
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 144 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0A6dqZySjyhhQ2ey:-KWvxMk4X0G1tILA9tAS:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L144)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> 'Parameter'
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl, line 1 at r6](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWvzDImxFpP4Dnl9vzE:-KWvzDImxFpP4Dnl9vzF:b-r95lom) ([raw file](https://github.com/powershell/xcomputermanagement/blob/2f1ece5a4032fe938d98d0d5dfc0e54da044945c/DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl#L1)):*
> ```
> [Description("This resource is used to activate a power plan.") : Amended,AMENDMENT, LOCALE("MS_409")]
> ```

@mbreakey3 This was a new file type for me, so don't know if this one is correct. How do I verify it?

---

*[Examples/Sample_xPowerPlan.ps1, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0qtvFOfll8ofqZmq:-KWvy_iKmWaEuhV6lOZ9:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Examples/Sample_xPowerPlan.ps1#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Can you add a description comment at the top about what this example does?
</details>
Done

---

*[Examples/Sample_xPowerPlan.ps1, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0xpItcV90j0y8T82:-KWvyaSrQSgxB5-f3-Sk:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Examples/Sample_xPowerPlan.ps1#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> 'Configuration'
</details>
Done

---

*[Examples/Sample_xPowerPlan.ps1, line 5 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0nejTrtMYO1mIrLY:-KWvybWSqV92n9B_6XS1:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Examples/Sample_xPowerPlan.ps1#L5)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> 'String[]'  
</details>
Done

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 23 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt1ClQ25PD9tkIcW9o:-KWvyd7tkLFGsajIuysP:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L23)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> you can remove this.
</details>
Done

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 34 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt1FjaofyWHGKUZ-2a:-KWvyeRSDFjiku2NrOVJ:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L34)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> remove this. And the extra newlines around it.
</details>
Done

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 82 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt25jKUU9GYogLfcfG:-KWvygFo3Rs0ShgNI66w:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L82)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> this should be checking that the correct error is thrown.
</details>
Done

---

*[Tests/Unit/MSFT_xPowerPlan.Tests.ps1, line 131 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt2GOJCiuU0hAhwOzz:-KWvyj6R08BGINX-RVD3:b-igzvrs) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/Tests/Unit/MSFT_xPowerPlan.Tests.ps1#L131)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> again, should be checking that the correct error is thrown
</details>
Done. Also added additional test to get 100% coverage after changing error handling.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KWvyvy17Ip5iYCSprVr:b-2udq7e)*
<!-- Sent from Reviewable.io -->

johlju commented:
Should I suppress this rule in Set-TargetResource?

```
Flagged PSSA rule(s) did not pass.
The following PSScriptAnalyzer errors need to be fixed or approved to be suppressed:
MSFT_xPowerPlan.psm1 (Line 88): 'Set-TargetResource' calls ShouldProcess/ShouldContinue but does not have the ShouldProcess attribute.
```

---

Review status: 7 of 12 files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KWw6lDjCGicDZrZ8hoB:bh8eaey)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
I've just recently (in the last 48 hours) had to suppress lots of instances of this rule violation (a change was made to DSCResource.tests that bought these to light). But I think this rule is OK to suppress for Set-TargetResource (only) because:
1. It is suppressed in PSDscResources (https://github.com/PowerShell/PSDscResources/blob/dev/DscResources/MSFT_UserResource/MSFT_UserResource.psm1#L89)
2. It is in the "Flag" list (https://github.com/PowerShell/DscResources/blob/master/PSSARuleSeverities.md#flag---occasionally-can-be-overruled-with-approval).

@KwirkyKat - this rule violation is popping up all over the repos - is it OK if we just blanket suppress it for ```Set-TargetResoruce``` only (no other functions) - rather than confirming for each repo?

---

Review status: 7 of 12 files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KWyKcaJU2QSYDd3AbjS:b-puwfja)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Reviewed 1 of 8 files at r3, 2 of 6 files at r6, 4 of 5 files at r7.
Review status: 11 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KX7RnwiqqBbsxZceTil:bmnf32a) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
> ```
> Subproject commit 987ae8c9dc3f1449a0be8dab7f77d04b097c2da3
> ```

Why is this line in DSCResource.Tests?

---

*[DSCResources/CommonResourceHelper.psm1, line 9 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7S-RSza4kxuaY8Z9P:-KX7S-RSza4kxuaY8Z9Q:bkadwf7) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResources/CommonResourceHelper.psm1#L9)):*
> ```PowerShell
>         The name of the resource as it appears before '.strings.psd1' of the localized string file.
>         For example:
>             WindowsOptionalFeature: MSFT_WindowsOptionalFeature
> ```

Change these examples to resources in xSqlServer

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 11 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KX7cr-QMgkSTGniOY4v:bsxfc2r) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Why is this line in DSCResource.Tests?
</details>
Hmm... Strange. I accidentally committed that in one of the commits. I tried to remove it, but apparently I did not do a good job about it. I will try to get this removed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Regarding the PSSA rule that Set-method is calling ShouldProcess/ShouldContinue. I can't see it doing that. Not even thru a helper function. So why is it complaining? 

---

Reviewed 1 of 8 files at r3, 1 of 3 files at r4, 4 of 5 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResources/CommonResourceHelper.psm1, line 9 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7S-RSza4kxuaY8Z9P:-KX7hlYd1s8t812-9dbb:b-340qda) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResources/CommonResourceHelper.psm1#L9)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Change these examples to resources in xSqlServer
</details>
Did you mean like this? If so, done. :)

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KX7fQXLGhGOEp74A5Jn:b-ffjeu5) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Hmm... Strange. I accidentally committed that in one of the commits. I tried to remove it, but apparently I did not do a good job about it. I will try to get this removed. 
</details>
The folder should be gone now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KX7foVDTtjWZuNQ_3LO:b-n98i5w)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KX7qqtLmKjQ_RCsIwzE:bhibbyw) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> The folder should be gone now
</details>
Great, also, what was the reason for deleting  the .gitignore file?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Reviewed 2 of 6 files at r6, 1 of 5 files at r7.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0fvyxVNrrgLx5gIU:-KX81lsqiQHsIIC3HwuS:bp0y2dw) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> I have to ask... Are you sure? :smile:
> `Test-xDscSchema` does not like that. And neither does `mofcomp.exe`.
> 
> ```
> PS > Test-xDscSchema .\MSFT_xPowerPlan.schema.mof
> .\MSFT_xPowerPlan.schema.mof (1): error SYNTAX 0X80044007: Illegal constant value. (Numeric value out of range or strings without quotes)
> Compiler returned error 0x80044007
> ```
</details>
Yeah, I guess this is fine. I know the mofs can be temperamental 

---

*[DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl, line 5 at r8](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX826LawP_tFFPNyDOl:-KX826LawP_tFFPNyDOm:bauz4sx) ([raw file](https://github.com/powershell/xcomputermanagement/blob/aa7e5752975218cda005b2c0f3baa1ae90aedc27/DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl#L5)):*
> ```
> {
>     [Key, Description("Specifies the resource is a single instance, the value must be 'Yes'") : Amended] String IsSingleInstance;
>     [Required, Description("The name of the power plan to activate.") : Amended] String Name;
> ```

I don't think 'Required, ' needs to be here, but other than that it looks good.

---

*[DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.strings.psd1, line 8 at r8](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX82bQfxZNWUS6eDUFF:-KX82bQfxZNWUS6eDUFG:b-3rtvfi) ([raw file](https://github.com/powershell/xcomputermanagement/blob/aa7e5752975218cda005b2c0f3baa1ae90aedc27/DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.strings.psd1#L8)):*
> ```PowerShell
>     PowerPlanNotFound = Unable to find the power plan '{0}'.
>     PowerPlanIsBeingActivated = Activating power plan '{0}'
>     PowerPlanIsBeingValidate = Validating power plan '{0}'
> ```

'PowerPlanIsBeingValidated'

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 8 files at r3, 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[DSCResources/CommonResourceHelper.psm1, line 9 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7S-RSza4kxuaY8Z9P:-KXCF_mAkg-LuXEdUKsU:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResources/CommonResourceHelper.psm1#L9)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Did you mean like this? If so, done. :)
</details>
Done.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8WsAjGpVct8BblBUd:-KXCFcP9S_ssqmhXXvKT:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L18)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Change it to a single-instance resource
</details>
Done.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 112 at r1](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KS8Xh_OKCyIydXLFhaI:-KXCFo_5TGKHpXmUysSm:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/1fc9e2bc21b9aab7d2276fdaa527e517c31a7ad0/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L112)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Done
</details>
Done.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 10 at r3](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KTPcy2Z7mU44sbpsP92-r3-10:-KXCFnddsUCkIbr5u272:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/74aebb6353c072fc2ffdd390da8514611ce08d0b/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L10)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Done
</details>
Done.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof, line 1 at r5](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWt0fvyxVNrrgLx5gIU:-KXCFsx3lYX_2gsZfB1U:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/eb02481712fd69f9c7c257136c1b15049dc7e757/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.schema.mof#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Yeah, I guess this is fine. I know the mofs can be temperamental 
</details>
Done.

---

*[DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl, line 1 at r6](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KWvzDImxFpP4Dnl9vzE:-KXCFgP1lpT2HOlOydd_:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/2f1ece5a4032fe938d98d0d5dfc0e54da044945c/DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> @mbreakey3 This was a new file type for me, so don't know if this one is correct. How do I verify it?
</details>
Done.

---

*[DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl, line 5 at r8](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX826LawP_tFFPNyDOl:-KXCFQha8rqwSAaHTy1h:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/aa7e5752975218cda005b2c0f3baa1ae90aedc27/DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.schema.mfl#L5)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> I don't think 'Required, ' needs to be here, but other than that it looks good.
</details>
Done

---

*[DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.strings.psd1, line 8 at r8](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX82bQfxZNWUS6eDUFF:-KXCF32AbBG8ernq74Ww:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/aa7e5752975218cda005b2c0f3baa1ae90aedc27/DSCResources/MSFT_xPowerPlan/en-US/MSFT_xPowerPlan.strings.psd1#L8)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> 'PowerPlanIsBeingValidated'
</details>
Done

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KXCFjr2FUA1XScwZ5jY:bjcm0tm) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Great, also, what was the reason for deleting  the .gitignore file?
</details>
Done. I thought we did not want git stuff in the repo. I added it back. 

---

*[DSCResources/xPowerPlan/xPowerPlan.schema.mof, line 4 at r2](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KRxm5t4TzM-giV7ZlE7-r2-4:-KXCFmA341kRtp0JNi54:b-896fix) ([raw file](https://github.com/powershell/xcomputermanagement/blob/98fed21fc48bbedd6771531fdc1bd4f40f3ef7df/DSCResources/xPowerPlan/xPowerPlan.schema.mof#L4)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> @TravisEz13 Requested changes done. If you would be so kind to review again whenever you have an opportunity
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
@mbreakey3 If you would be so kind to look thru this again 😄

mbreakey3 commented:
Reviewed 1 of 8 files at r3, 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KXCXGudhgoPJdcHvwsN:bmb2ln6) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Done. I thought we did not want git stuff in the repo. I added it back. 
</details>
Yeah, we need it to make sure DSCResource.Tests doesn't get included.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 53 at r9](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KXCXfw1SuCl20uMfdrP:-KXCXfw1SuCl20uMfdrQ:b-fo2j6j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/c28435c62a58e0c01649f4f74117df9b606b65de/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L53)):*
> ```PowerShell
>     if ($plan)
>     {
>         if( $plan.IsActive )
> ```

```powershell
if ($plan.IsActive)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 8 files at r3, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1, line 53 at r9](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KXCXfw1SuCl20uMfdrP:-KXCZSTmUPVA3UsYT_yT:bb6un1j) ([raw file](https://github.com/powershell/xcomputermanagement/blob/c28435c62a58e0c01649f4f74117df9b606b65de/DSCResources/MSFT_xPowerPlan/MSFT_xPowerPlan.psm1#L53)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> ```powershell
> if ($plan.IsActive)
> ```
</details>
Done

---

*[DSCResource.Tests, line 1 at r7](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-KX7RnwiqqBbsxZceTik:-KXCZUtOwOaqUu1LJfEU:bu830vo) ([raw file](https://github.com/powershell/xcomputermanagement/blob/fa445011054b9ee3980c15a724884c89ae0b313e/DSCResource.Tests#L1)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Yeah, we need it to make sure DSCResource.Tests doesn't get included.
</details>
Right!
Done. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r3, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcomputermanagement/47#-:-KXC_PwyhOrG0VcEOdS8:bnfp4nl)*
<!-- Sent from Reviewable.io -->
